docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,804 @@
+    return pv.pubKey, nil
+}
+
+// SignVote signs a canonical representation of the vote. If signExtension is true, it also signs the vote extension.
+func (pv *CryptoProviderPV) SignVote(chainID string, vote *Vote, signExtension bool) error {
+    signer := pv.provider.GetSigner()
+
+    // code for getting voteBytes goes here
+    // voteBytes := ...
+
+    // The underlying signer needs these parameters so we pass them through SignerOptions
+    options := SignerOptions{
+        "chainID": chainID,
+        "vote":    vote,
+    }
+ 
+    sig, _ := signer.Sign(voteBytes, options)
+    vote.Signature = sig
+    return nil
+}
+
+// SignProposal signs a canonical representation of the proposal
+func (pv *CryptoProviderPV) SignProposal(chainID string, proposal *Proposal) error {
+    signer := pv.provider.GetSigner()
+
+    // code for getting proposalBytes goes here
+    // proposalBytes := ...
+
+    // The underlying signer needs these parameters so we pass them through SignerOptions
+    options := SignerOptions{
+        "chainID":  chainID,
+        "proposal": proposal,
+    }
+ 
+    sig, _ := signer.Sign(proposalBytes, options)
+    proposal.Signature = sig
+    return nil
+}
+
+// SignBytes signs an arbitrary array of bytes
+func (pv *CryptoProviderPV) SignBytes(bytes []byte) ([]byte, error) {
+    signer := pv.provider.GetSigner()
+    return signer.Sign(bytes, SignerOptions{})
+}
+
+```
+
+*Note:* Each provider (File, Socket, SignerClient, RetrySignerClient) will need to be modified to satisfy the `CryptoProvider` interface.
+
+
+#### Loading and storing implementations
+
+Alternatives:
+
+* Low impact / minimal dependencies: The corresponding CryptoProvider [Protobuf message](#protobuf-message-structure) can be directly stored on disk in a dedicated directory in an encoding format of choice (text, JSON).
+
+* Greater impact / better security: Use cosmos-sdk's [Keyring](https://github.com/cosmos/cosmos-sdk/blob/439f2f9d5b5884bc9df4b58d702555330549a898/crypto/keyring/keyring). to manage CryptoProviders along with its private keys. This specifically applies to the `FilePV` implementation, which could store its private keys in Keyring instead of a file in the filesystem. This approach will require decoupling the Keyring package from the cosmos-sdk to avoid having CometBFT import artifacts from that repository.
+
+
+Implementations of crypto providers (previously `Privval` implementations) should be in their own directory:
+
+```
+cometbft/
+├── privval/
+│   ├── providers/
+│   │   ├── file_provider.go
+│   │   ├── signer_provider.go
+│   │   ├── retry_signer_provider.go
+│   │   ├── mock_provider.go
+├── types/
+│   ├── priv_validator.go
+```
+
+#### Other considerations
+
+##### Node startup
+
+Node's configuration file should include the name of the `CryptoProvider` to be loaded at startup. Also, the startup logic will need to be changed from creating a validator to loading a `CryptoProvider`
+
```

Comment:
First of all, thanks a lot for putting this together 🙏.
Two general remarks on the CometBFT section.
* Before we start implementing this on CometBFT, we need to have and ADR in CometBFT repo with this information. That ADR can refer to this one for basic concepts, such as the crypto provider interface, etc. But the changes intended for CometBFT must be in a CometBFT ADR as some point.
* I am missing an important part of what I had understood this work is about: adding new curves and new (remote?) signers. I know you discuss it in the general sections, but I would like to see how it would come together in CometBFT. How would that work?

Reply:
- Sure, I think the whole section referring on CometBFT's implementation can be extracted and inserted into a new "standalone" ADR and leave this one for just the sdk (adding the proper links and references, of course). Do you think we can do that PR now?
- CometBFT's codebase I think is a "lighter" case than the sdk, nevertheless the benefits are the same in both: having the ability to have "pluggable cryptography tools" thus enhancing modularity without much changes in the codebase.